### PR TITLE
feat(File Upload): Add retry and upload a new file workflow under feature flag

### PR DESCRIPTION
**Pull request by dat-gcio**
## Description
Implement new workflow to reduce the number of actions veterans have to take in order to retry, or upload a new file when an error happens.

## Testing done
- [ ] unit
- [ ] manual

## Screenshots

File type not allowed:
image

File too large:
image

File too small: 
image

Network error:
image

Don't allow additional upload if error present:
Before:
image

Now: 
image

## Acceptance criteria
- [ ] Update connection error:
We're sorry. We had a connection problem. Please try again.
[Try again]  [Cancel]

- [ ] Update File too large error:
We couldn't upload your file because it's too big. Please make sure the file is 10MB or less and try again.
[Upload a new file] [Cancel]

- [ ] Update File too small error:
We couldn't upload your file because it's too small. Please make sure the file is 1B or more and try again.
[Upload a new file] [Cancel]

- [ ] Update Wrong file type error
We couldn't upload your file because we can't accept this type of file. Please make sure the file is a .pdf, .jpeg, jpg, or .png file and try again.
[Upload a new file] [Cancel]

- [ ] Forms with multiple file upload option should remove "Add another button" on error
- [ ] Try again should resend the file to the backend
- [ ] Upload another file should delete the file, and prompt for file to upload

## Definition of done
- [ticket](https://github.com/department-of-veterans-affairs/va.gov-team/issues/26419)

### CODE REVIEW

**Comment by dat-gcio** [comment]
Not yet ready for review. Accidentally clicked.

**Comment by dat-gcio** [comment]
> Not yet ready for review. Accidentally clicked. Meant to open review instance link.
PR ready for review

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx]: Why are uploads being allowed if there are errors?

**Review by bkjohnson — COMMENTED** [review]
This is a first round - I'll come back again and leave some more comments/questions.

File: src/platform/forms-system/src/js/actions.js
Comment (dat-gcio) [reply src/platform/forms-system/src/js/actions.js:364]: The 'Please delete this file' and try again verbiage was added in the last iteration. It is to provide users with direct directions. It is helpful keep it, when the new button/workflow is not enabled.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx:416]: Ah, that makes sense. I'll keep it in mind for next time.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx]: It currently allows it, so I wrapped this 'enhancement' under the feature flag.

image

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx:195]: Good call. Added.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx]: Ended up removing this change, since it doesn't need guarding.  The only time we need the 'file' object is on resending and the action.js function takes care of sending the file back anyways.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (bkjohnson) [inline src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:1156]: Naming.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (bkjohnson) [inline src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:1206]: Same comment here about the selector - I'm not sure what it's referring to.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (bkjohnson) [inline src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:1253]: Naming.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (bkjohnson) [inline src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:1294]: Naming.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (bkjohnson) [inline src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:1341]: It's not obvious from this that the "retry" button is a primary button - can you make that more clear?

**Review by bkjohnson — CHANGES_REQUESTED** [review]
In the tests for `FileField`, I think it would be good to move some of the `const`s that don't change between tests (`idSchema`, `schema`) up to the `describe` level rather than redefine them in each `it` block.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx:448]: I needed the actual file input in the DOM even when the main file 'upload' button is not present. It allows the 'Retry' button to work properly.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx]: You are right, it is out of scope. Opted to remove it, and do it in another PR linking to that ticket.

File: src/platform/forms-system/test/js/actions.unit.spec.js
Comment (dat-gcio) [reply src/platform/forms-system/test/js/actions.unit.spec.js]: Updated with your suggestion

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (dat-gcio) [reply src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx]: I am targeting the label that wraps that span. I renamed the idSchema.$id prop in the test to help identify it. Let me if that is not what you meant.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (dat-gcio) [reply src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx]: Added a comment to help clarify. Let me know if you meant something else.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (dat-gcio) [reply src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx]: Updated name, and added comment.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (dat-gcio) [reply src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx]: Updated name, and added comment.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (dat-gcio) [reply src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx]: Updated name, and added comment.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (dat-gcio) [reply src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx]: Added a comment to help discern that the id is interpolated.

File: src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx
Comment (dat-gcio) [reply src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx]: Updated and added comment

**Comment by dat-gcio** [comment]
> In the tests for `FileField`, I think it would be good to move some of the `const`s that don't change between tests (`idSchema`, `schema`) up to the `describe` level rather than redefine them in each `it` block.

Thanks for the suggestion. Moved the reused `const`s up to the `describe` level. Also updated the test description to be more descriptive and added comments above the button queries to describe the target. I was too caught up in handling all the cases of the feature flag, and should have done a better job with the test writing itself. Sorry you had to go through all of that. 
image

**Comment by timwright12** [comment]
Don't have a lot of context into these updates, so I'll defer to @bkjohnson for PR approval. Let me know if you do need me though.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [inline src/platform/forms-system/src/js/fields/FileField.jsx:267]: Does the condition after the `||` mean that this would allow uploads when there _aren't_ errors? I see the `!` has been added to the beginning of `files`.

Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx:267]: Yup. If the flag is on, only allow additional uploads when there are no errors present.

**Comment by dat-gcio** [comment]
Hey @bkjohnson, do you think this can be approved by 2PM EST today? Need to have it in staging for staging review.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [reply src/platform/forms-system/src/js/fields/FileField.jsx:267]: Does the `WhenError` part of the name still make sense?

Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx:267]: Because the feature flag is mixed in there, it does. The variable will need a rename when the flag is removed. With the flag off (`!enableShortWorkflow`) we `allowUploadWhenError`. When the flag is on, we don't `allowUploadWhenError`.

**Comment by dat-gcio** [comment]
> Hey @bkjohnson, do you think this can be approved by 2PM EST today? Need to have it in staging for staging review.

Staging review has been pushed back a week, so we have until Thursday to get this in.

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (bkjohnson) [reply src/platform/forms-system/src/js/fields/FileField.jsx:267]: I think this should be renamed to reflect what the value represents regardless of whether or not the feature flag is flipped (`allowUpload` possibly), and a comment added explaining how the feature flag changes the criteria for allowing an upload.

This is defined on line 258 and not used until line 450. If a future person is reading this code and sees `allowUploadWhenError && ...`, there's a good chance that they may not be thinking "but invert this if a specific feature flag is on", because the definition won't be visible. Even if the feature flag is removed and becomes the default a week from now, I'd rather avoid some potential awkwardness with the naming.

**Review by bkjohnson — COMMENTED** [review]
Hey, sorry this missed staging review. I don't mean to be too nitpicky, but the forms library already has some problems with clarity/complexity, and I'm trying to make sure that others who read the code will not have too much difficulty in understanding what's going on.

The nature of the forms library also means that making small PRs can be hard or impossible, and longer PRs simply take more time to review. If a PR _is_ long, leaving your own review can help speed things up when another codeowner steps in to review it.

Thanks for your patience!

File: src/platform/forms-system/src/js/fields/FileField.jsx
Comment (dat-gcio) [reply src/platform/forms-system/src/js/fields/FileField.jsx]: Ok, finally understood what you were telling me. Name changed and comment added. And no worries on not making staging review. I don't think anyone on my team knew we had to get the artifact in staging a week before the staging review date. We all thought it was by the date of the staging review. 

And I really appreciate your thoroughness, especially since I am very new to va.gov.

**Review by bkjohnson — APPROVED** [review]
:tada:

**Review by timwright12 — APPROVED** [review]

**Pull request merged**
